manifests/kustomize/base/installs/multi-user/pipelines-profile-controller/sync.py

```diff
@@ -281,5 +310,81 @@ def do_POST(self):
+        """
+        Returns a dict of settings from environment variables relevant to the controller
+
+        Environment settings can be overridden by passing them here as arguments.
+
+        Settings are pulled from the all-caps version of the setting name.  The
+        following defaults are used if those environment variables are not set
+        to enable backwards compatability with previous versions of this script:
+            visualization_server_image: gcr.io/ml-pipeline/visualization-server
+            visualization_server_tag: value of KFP_VERSION environment variable
+            frontend_image: gcr.io/ml-pipeline/frontend
+            frontend_tag: value of KFP_VERSION environment variable
+            disable_istio_sidecar: Required (no default)
+            mlpipeline_minio_access_key: Required (no default)
+            mlpipeline_minio_secret_key: Required (no default)
+        """
+        settings = {}
+        settings["visualization_server_image"] = \
+            visualization_server_image or \
+            os.environ.get("VISUALIZATION_SERVER_IMAGE", "gcr.io/ml-pipeline/visualization-server")
+
+        settings["frontend_image"] = \
+            frontend_image or \
+            os.environ.get("FRONTEND_IMAGE", "gcr.io/ml-pipeline/frontend")
+
+        # Look for specific tags for each image first, falling back to
+        # previously used KFP_VERSION environment variable for backwards
+        # compatibility
+        settings["visualization_server_tag"] = \
+            visualization_server_image or \
+            os.environ.get("VISUALIZATION_SERVER_TAG") or \
+            os.environ["KFP_VERSION"]
+
+        settings["frontend_tag"] = \
+            visualization_server_image or \
+            os.environ.get("FRONTEND_TAG") or \
+            os.environ["KFP_VERSION"]
+
+        settings["disable_istio_sidecar"] = \
+            disable_istio_sidecar if disable_istio_sidecar is not None \
+            else os.environ.get("DISABLE_ISTIO_SIDECAR") == "true"
+
+        settings["mlpipeline_minio_access_key"] = \
+            mlpipeline_minio_access_key or \
+            base64.b64encode(bytes(os.environ.get("MINIO_ACCESS_KEY"), 'utf-8')).decode('utf-8')
+
+        settings["mlpipeline_minio_secret_key"] = \
+            mlpipeline_minio_access_key or \
+            base64.b64encode(bytes(os.environ.get("MINIO_SECRET_KEY"), 'utf-8')).decode('utf-8')
+
+        return settings
+
+    @classmethod
+    def partial_init_from_environment(cls):
+        """
+        Returns this class partially instantiated with settings inferred from environment variables
+
+        Useful for building a version of this Controller that can be consumed by
+        HTTPServer as a handler, but that includes all settings inferred from
+        environment variables, an a testable way.
+        """
+        settings = cls.get_settings_from_environment()
+        return partial(cls, **settings)
```

Comment:
Hi @ca-scribner! The improvements & testing are awesome. I only have a minor comment here:

This might be my lack of experience of python, what does this partial method mean? Is it a well known language construct?
One of the goals for this script is to be as simple as possible, because we can imagine there can be very ad-hoc requirements when using multi-user KFP in a production environment.

I wonder if we can make the script testable without using advanced language features, what do you think about:
```python
def main():
  settings = get_settings_from_env()
  serve_controller_forever(**settings)

def get_settings_from_env():
  ...
  return settings

def serve_controller_forever(port, kfp_version, ...):
  class Controller(BaseHTTPRequestHandler):
    ...
  HTTPServer(("", port), Controller).serve_forever()

if __name__ == "__main__":
  main()
```

Reply:
Thanks @Bobgy, great feedback!  Agreed it felt more complex than needed.  Your suggestions are very helpful.

Partial is probably in the intermediate range of features?  Definitely not known by everyone though.  It lets you partly fill out the arguments for a function, with the returned object acting like the original function except needing only the un-filled arguments.  So given:
```
def f(a, b):
  return a + b
```
these would be equivalent:
```
x = f(1, 2)  # result is x=3
```

```
f_ = partial(f, b=2)
x = f_(1)  # result is x=3
```

It is sort of a low-code way to make a factory.

How do you feel about:
```
def main():
    settings = get_settings_from_environment()
    server = server_factory(**settings)
    server.serve_forever()

def get_settings_from_env():
    ...
    return settings

def server_factory(port, kfp_version, ...):
    class Controller(BaseHTTPRequestHandler):
    ...
    server = HTTPServer(("", port), Controller)
    return server


def serve_controller_forever(port, kfp_version, ...):
    class Controller(BaseHTTPRequestHandler):
        ...
  
    return HTTPServer(("", port), Controller).serve_forever()

if __name__ == "__main__":
  main()
```

This lets us create servers without automatically calling `.serve_forever()` (good for testing), but more explicitly calls out that we're building bespoke class objects based on the passed arguments.  I've fleshed it out more in a commit I'll push now 

Feedback welcome!